File: cli/core/formats/sbv.py
"""YouTube SBV parser and rebuilder."""

from __future__ import annotations

import re

from ..srt_parser import SubtitleBlock, normalize_text
from .types import SubtitleDocument

_SBV_LINE_RE = re.compile(
    r"^(\d+):(\d{2}):(\d{2})\.(\d{3})\s*,\s*"
    r"(\d+):(\d{2}):(\d{2})\.(\d{3})$"
)


def parse_sbv(content: str) -> SubtitleDocument:
    text = normalize_text(content).strip()
    chunks = re.split(r"\n\n+", text)

    blocks: list[SubtitleBlock] = []
    original_ts: list[str] = []
    n = 1

    for chunk in chunks:
        lines = chunk.split("\n")
        m = _SBV_LINE_RE.match(lines[0].strip())
        if not m:
            continue
        g = m.groups()
        start = f"{int(g[0]):02d}:{g[1]}:{g[2]},{g[3]}"
        end = f"{int(g[4]):02d}:{g[5]}:{g[6]},{g[7]}"
        cue_text = normalize_text("\n".join(lines[1:]))

        blocks.append(
            SubtitleBlock(
                number=n, timestamp=f"{start} --> {end}", text=cue_text
            )
        )
        original_ts.append(lines[0].strip())
        n += 1

    def rebuild(translated: list[SubtitleBlock]) -> str:
        # Keyed by block number, not list position: a short or reordered
        # translation must not shift every cue onto the wrong timestamp.
        text_by_number = {b.number: b.text for b in translated}
        out = [
            f"{ts}\n{text_by_number.get(src.number, src.text)}"
            for ts, src in zip(original_ts, blocks, strict=True)
        ]
        return "\n\n".join(out) + "\n"

    return SubtitleDocument(format="sbv", blocks=blocks, rebuild=rebuild)

File: web/src/app/core/subtitle-formats/sbv.ts
import { SubtitleBlock } from '../srt-parser';
import { SubtitleDocument, normalizeNewlines, pad2, stripBom } from './types';

/** YouTube SBV: `H:MM:SS.mmm,H:MM:SS.mmm\ntext\n\n`, no block numbers. */

const SBV_LINE_RE =
  /^(\d+):(\d{2}):(\d{2})\.(\d{3})\s*,\s*(\d+):(\d{2}):(\d{2})\.(\d{3})$/;

export function parseSbv(content: string): SubtitleDocument {
  const text = normalizeNewlines(stripBom(content)).trim();
  const chunks = text.split(/\n\n+/);

  const blocks: SubtitleBlock[] = [];
  const originalTsLines: string[] = [];
  let n = 1;

  for (const chunk of chunks) {
    const lines = chunk.split('\n');
    const m = lines[0].trim().match(SBV_LINE_RE);
    if (!m) continue;

    const start = `${pad2(Number(m[1]))}:${m[2]}:${m[3]},${m[4]}`;
    const end = `${pad2(Number(m[5]))}:${m[6]}:${m[7]},${m[8]}`;
    const cueText = lines.slice(1).join('\n');

    blocks.push({
      number: n++,
      timestamp: `${start} --> ${end}`,
      text: cueText,
    });
    originalTsLines.push(lines[0].trim());
  }

  return {
    format: 'sbv',
    blocks,
    rebuild(translated) {
      return translated
        .map((b, i) => `${originalTsLines[i] ?? srtToSbv(b.timestamp)}\n${b.text}`)
        .join('\n\n') + '\n';
    },
  };
}

function srtToSbv(srt: string): string {
  const m = srt.match(
    /(\d{2}):(\d{2}):(\d{2}),(\d{3})\s*-->\s*(\d{2}):(\d{2}):(\d{2}),(\d{3})/,
  );
  if (!m) return srt;
  const stripLeadingZeroHour = (h: string) => (h === '00' ? '0' : String(Number(h)));
  return `${stripLeadingZeroHour(m[1])}:${m[2]}:${m[3]}.${m[4]},${stripLeadingZeroHour(m[5])}:${m[6]}:${m[7]}.${m[8]}`;
}
